pkg/requests/util/util.go

```diff
@@ -3,13 +3,23 @@ package util
 import (
 	"net/http"
 
+	"github.com/google/uuid"
 	middlewareapi "github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/middleware"
 )
 
+const (
+	XForwardedProto = "X-Forwarded-Proto"
+	XForwardedHost  = "X-Forwarded-Host"
+	XForwardedURI   = "X-Forwarded-Uri"
+	XRequestID      = "X-Request-Id"
+)
+
+var requestIDHeader = XRequestID
```

Comment:
Does this get overriden when the config is loaded? Not a huge fan of this :thinking: 

How many times is this actually used in the code? I see it's used in tests a lot but this should actually only be in the scope middleware right? Maybe logging? Perhaps we want to pass in the configured value every time?

I guess this is kind of like how things are done in the logger package right, configuring the default logger? Perhaps lets add a comment explaining why this is done this way and leave it?

Reply:
So `GetRequestID` is a little overloaded 😅 

```
func GetRequestID(req *http.Request) string {
	scope := middlewareapi.GetRequestScope(req)
	if scope != nil {
		return scope.RequestID
	}
	rid := req.Header.Get(requestIDHeader)
	if rid != "" {
		return rid
	}
	return uuid.New().String()
}
```

The first time it is called for a request, it is in the `Scope` initialization itself, so it falls through to the 2nd or 3rd case and sets the scope (only time our configured `requestIDHeader` is used). Every other time in the lifecycle, scope is set so it is pulling from that first block.

I guess we are technically memoizing on the `middlewareapi.Scope` for how we consume it in practice.

Reply:
Should we separate this out, so we have two different things.

1. In the request scope middleware, we add or generate the request ID.
2. We have a helper here to read it from the scope

This way we can pass the request ID header as a parameter when we construct the middleware scope, will likely be cleaner than the current implementation right?

Reply:
Question for you since you are more familiar with the ordering of everything:

Since `GetRequestID` is used in logging/error pages, I opted to merge the two cases (rather than `GetRequestID` always assuming a `scope` was already initialized) in case we had a scenario where we errored or attemtped to log before the scope was initialied.

That's the benefit of keeping aligned. But Is that a thing? Otherwise you are right, could be split to be cleaner.

Reply:
The scope middleware should be the first thing in the request chain, so, unless we do any logging that would use the request ID as part of the middleware scope, then we are safe